Sources/NIOOpenSSL/SSLPKCS12Bundle.swift

```diff
@@ -0,0 +1,218 @@
+    ///
+    /// - parameters:
+    ///     - file: The path to the PKCS#12 bundle on disk.
+    ///     - passphrase: The passphrase used for the bundle, as a sequence of UTF-8 bytes.
+    public init<Bytes: Collection>(file: String, passphrase: Bytes?) throws where Bytes.Element == UInt8 {
+        guard openSSLIsInitialized else { fatalError("Failed to initialize OpenSSL") }
+
+        let fileObject = fopen(file, "rb")
+        defer {
+            fclose(fileObject)
+        }
+
+        let p12 = d2i_PKCS12_fp(fileObject, nil)
+        defer {
+            p12.map(PKCS12_free)
+        }
+
+        if let p12 = p12 {
+            try self.init(ref: p12, passphrase: passphrase)
+        } else {
+            throw OpenSSLError.unknownError(.init())
+        }
+    }
+
+    /// Create a `OpenSSLPKCS12Bundle` from the given bytes on disk,
+    /// assuming it has no passphrase.
+    ///
+    /// If the bundle does have a passphrase, call `init(file:passphrase:)` instead.
+    ///
+    /// - parameters:
+    ///     - file: The path to the PKCS#12 bundle on disk.
+    public init(file: String) throws {
+        try self.init(file: file, passphrase: nil as [UInt8]?)
+    }
+
+    /// Create a `OpenSSLPKCS12Bundle` from the given bytes in memory,
+    /// assuming it has no passphrase.
+    ///
+    /// If the bundle does have a passphrase, call `init(buffer:passphrase:)` instead.
+    ///
+    /// - parameters:
+    ///     - buffer: The bytes of the PKCS#12 bundle.
+    public init(buffer: [UInt8]) throws {
+        try self.init(buffer: buffer, passphrase: nil as [UInt8]?)
+    }
+}
+
+
+internal extension Collection where Element == UInt8 {
+    /// Provides a contiguous copy of the bytes of this collection in a heap-allocated
+    /// memory region that is locked into memory (that is, which can never be backed by a file),
+    /// and which will be scrubbed and freed after use.
+    ///
+    /// This method should be used when it is necessary to take a secure copy of a collection of
+    /// bytes. Its implementation relies on OpenSSL directly.
+    internal func withSecureBytes<T>(_ block: (UnsafePointer<Int8>) throws -> T) throws -> T {
+        // We need to allocate some memory and prevent it being swapped to disk while we use it.
+        // For that reason we use mlock.
+        // Note that here we use UnsafePointer and UnsafeBufferPointer. Ideally we'd just use
+        // the buffer pointer, but 4.0 doesn't have the APIs we want: specifically, .allocate
+        // and .withMemoryRebound(to:) are only added to the buffer pointers in 4.1.
+        let bufferSize = Int(self.count) + 1
+        let ptr = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
+        let bufferPtr = UnsafeMutableBufferPointer(start: ptr, count: bufferSize)
+        defer {
+            ptr.deallocate()
+        }
+
+        if mlock(bufferPtr.baseAddress!, bufferPtr.count) != 0 {
+            throw IOError(errnoCode: errno, function: "mlock")
```

Comment:
this isn't guaranteed to give you the right `errno`. Can we move into the `Posix` module?

Reply:
ok, as of our chat a minute ago: this should then have its own `Posix` wrapper following `NIO`'s pattern

Reply:
I agree, but are you open to me doing this as a follow-up PR? There are a few other sys calls in the package, I'd like to fix it all in one go.

Reply:
@Lukasa yes!